Make tcp connections timeout after 1sec

canndrew opened the pull request:
Fixes https://github.com/nblockchain/TcpEchoSharp/issues/7

canndrew committed:
Make tcp connections timeout after 1sec

Fixes https://github.com/nblockchain/TcpEchoSharp/issues/7

knocte commented on src/Client/Client.cs:
```diff
@@ -31,7 +31,9 @@ public abstract class Client {
                 socket.ReceiveTimeout = 500;
 
-                await socket.ConnectAsync (endpoint, port);
+                if (!socket.ConnectAsync (endpoint, port).Wait (1000)) {
```
let's change the .ReceiveTimeout to be 1000 too, and we store it in a constant

canndrew commented:
For the test, it would make sense to connect to an address we know won't respond. Is there some way in C# we can open a TCP listening socket but not send back a SYN-ACK when our socket tries to connect? The only way I can think of is (on linux) to use network namespaces and tun/tap, since raw sockets require root/admin privileges.

canndrew commented:
Ping @c410-f3r

canndrew committed:
Add tcp timeout constant

Put tcp timeout in a constant and use it for the receive timeout aswell.

knocte commented:
Mmm, I think that using `.Wait` is a bad practice because it makes the thread block, thus ripping the advantages of using the async/await model. What we need is do something like this: https://stackoverflow.com/questions/4238345/asynchronously-wait-for-taskt-to-complete-with-timeout

canndrew committed:
make tcp timeout asynchronous

canndrew commented:
Right, of course. Fixed.

knocte committed:
Catch timeout exception to be wrapped in ConnectionUnsuccessfulException

knocte commented:
@canndrew this could have broken the already existing tests, so I added a new catch.

And I was going to merge the PR after this, however, I was curious to do a little experiment: remove the `break;` from the `ConnectWithElectrumServersTransactionGet` test, and the only effect should be that it takes a bit longer until it executes on all servers right? however, this PR seems to make it hang. Can you investigate? Putting some Console.WriteLine statements might help to easily find where it gets stuck.

knocte commented:
s/all tests/all servers/

knocte committed:
ClientTests: better logging

knocte committed:
Also handle timeouts after conn has been established

knocte committed:
Make test reach all servers, and count, not stop at 1st successful

We can do this now that we have proper timeout handling.
This way it will be easier to catch connection/transmission bugs.

knocte committed:
Refactor tests to follow same structure

knocte commented:
Nevermind, found the problem!
